sway/desktop/fx_renderer.c

```diff
@@ -371,27 +562,67 @@ struct fx_renderer *fx_renderer_create(struct wlr_egl *egl) {
 	return NULL;
 }
 
-void fx_renderer_begin(struct fx_renderer *renderer, uint32_t width, uint32_t height) {
-	// Create and render the stencil buffer
-	if (renderer->stencil_buffer_id == 0) {
-		glGenRenderbuffers(1, &renderer->stencil_buffer_id);
-	}
-	glBindRenderbuffer(GL_RENDERBUFFER, renderer->stencil_buffer_id);
-	glRenderbufferStorage(GL_RENDERBUFFER, GL_STENCIL_INDEX8, width, height);
-	glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_STENCIL_ATTACHMENT,
-			GL_RENDERBUFFER, renderer->stencil_buffer_id);
+void fx_renderer_fini(struct fx_renderer *renderer) {
+	release_fx_framebuffer(&renderer->main_buffer);
+	release_fx_framebuffer(&renderer->blur_buffer);
+	release_fx_framebuffer(&renderer->effects_buffer);
+	release_fx_framebuffer(&renderer->effects_buffer_swapped);
+	release_stencil_buffer(&renderer->stencil_buffer_id);
+}
 
-	glViewport(0, 0, width, height);
+void fx_renderer_begin(struct fx_renderer *renderer, struct sway_output *sway_output,
+		pixman_region32_t *original_damage) {
+	struct wlr_output *output = sway_output->wlr_output;
+
+	int width, height;
+	wlr_output_transformed_resolution(output, &width, &height);
+
+	renderer->sway_output = sway_output;
+	renderer->original_damage = original_damage;
+	// Store the wlr framebuffer
+	glGetIntegerv(GL_FRAMEBUFFER_BINDING, &renderer->wlr_fb);
+
+	// Create the main framebuffer
+	fx_create_framebuffer(output, &renderer->main_buffer, true);
+	// Create the stencil buffer and attach it to our main_buffer
+	create_stencil_buffer(output, &renderer->stencil_buffer_id);
+
+	// Create a new blur/effects framebuffers
+	fx_create_framebuffer(output, &renderer->effects_buffer, false);
+	fx_create_framebuffer(output, &renderer->effects_buffer_swapped, false);
 
 	// refresh projection matrix
 	matrix_projection(renderer->projection, width, height,
 		WL_OUTPUT_TRANSFORM_FLIPPED_180);
 
 	glBlendFunc(GL_ONE, GL_ONE_MINUS_SRC_ALPHA);
+
+	// Bind to our main framebuffer
+	fx_bind_framebuffer(&renderer->main_buffer, width, height);
 }
 
-void fx_renderer_end() {
-	// TODO
+void fx_renderer_end(struct fx_renderer *renderer) {
+	struct wlr_output *output = renderer->sway_output->wlr_output;
+
+	// Draw the contents of our buffer into the wlr buffer
+	glBindFramebuffer(GL_FRAMEBUFFER, renderer->wlr_fb);
+
+	float clear_color[] = {0.0f, 0.0f, 0.0f, 1.0f};
+	if (pixman_region32_not_empty(renderer->original_damage)) {
+		int nrects;
+		pixman_box32_t *rects = pixman_region32_rectangles(renderer->original_damage, &nrects);
+		for (int i = 0; i < nrects; ++i) {
+			scissor_output(output, &rects[i]);
+			fx_renderer_clear(clear_color);
+		}
+	}
+	fx_render_whole_output(renderer, renderer->original_damage,
+			&renderer->main_buffer.texture);
```

Comment:
This feels like it belongs in `render.c`. If we can move `fx_render_whole_output` to `render.c`, we can also move `scissor_output` back there.

Reply:
My reasoning there is twofold: firstly, it is only called when ending the rendering. Also, I'd really like to keep damage tracking in render.c, which will be really important once we support vulkan.

Reply:
If this has to occur only after drawing the buffer into wlr_fb though, I understand the limitations

Reply:
We could always move most of the code into render.c and leave the last two lines that releases the buffers?